### PR TITLE
CB-13110:Add the encryption Arn parameter passed from CLI to AWS env creation in Cloudbreak

### PR DESCRIPTION
**Pull request by satyabollineni**
Created a Environment with the encryption Key Arn from CLI and verified that the parameter encryption-key-arn is added to the DB:Screenshot 2021-07-29 at 5 21 34 PM

### CODE REVIEW

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/request/aws/AwsDiskEncryptionParameters.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+public class AwsDiskEncryptionParameters implements Serializable {
+
+    @VisibleForTesting
+    static final String AWS_ENCRYPTION_KEY_INVALID_MESSAGE = "The identifier of the AWS Key Management Service (AWS KMS) " +
+            "customer master key (CMK) to use for Amazon EBS encryption.\n" +
+            "You can specify the key ARN in the below format:\n" +
+            "Key ARN: arn:partition:service:region:account-id:resource-type/resource-id. " +
+            "For example, arn:aws:kms:us-east-1:012345678910:key/1234abcd-12ab-34cd-56ef-1234567890ab.\n";
+
+    @Pattern(regexp = "^arn:(aws|aws-cn|aws-us-gov):kms:[a-zA-Z0-9-]+:[0-9]+:key/[a-zA-Z0-9-]+$",
```
Comment (bergerdenes) [inline environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/request/aws/AwsDiskEncryptionParameters.java:23]: The key part is a GUID and this pattern does not validate against that fact

**Review by bergerdenes — CHANGES_REQUESTED** [review]

**Comment by doktoric** [comment]
Can you please one commit which contains all the change?

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/request/aws/AwsDiskEncryptionParameters.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import javax.validation.constraints.Pattern;
+import com.google.common.annotations.VisibleForTesting;
+
+import com.sequenceiq.environment.api.doc.environment.EnvironmentModelDescription;
+
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel(value = "AwsDiskEncryptionV1Parameters")
+public class AwsDiskEncryptionParameters implements Serializable {
```
Comment (lajosrodek) [inline environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/request/aws/AwsDiskEncryptionParameters.java:14]: Once #11024 gets merged, please pull & rebase so that these changes can disappear here. Furthermore, please be sure to squash your commits into a single one.

**Review by lajosrodek — COMMENTED** [review]
One general note.

File: structuredevent-model/src/test/java/com/sequenceiq/environment/parameter/dto/AwsParametersDtoTest.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package com.sequenceiq.environment.parameter.dto;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
```
Comment (bergerdenes) [inline structuredevent-model/src/test/java/com/sequenceiq/environment/parameter/dto/AwsParametersDtoTest.java:9]: You are not mocking anything here so please remove extension

File: structuredevent-model/src/test/java/com/sequenceiq/environment/parameter/dto/AwsDiskEncryptionParametersDtoTest.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
```
Comment (bergerdenes) [inline structuredevent-model/src/test/java/com/sequenceiq/environment/parameter/dto/AwsDiskEncryptionParametersDtoTest.java:11]: Mockito is not used

File: structuredevent-model/src/test/java/com/sequenceiq/environment/parameter/dto/AwsParametersDtoTest.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+public class AwsParametersDtoTest {
```
Comment (bergerdenes) [inline structuredevent-model/src/test/java/com/sequenceiq/environment/parameter/dto/AwsParametersDtoTest.java:10]: This test is not testing what you meant. You should test that `AwsParameterDto` indeed contains an `AwsDiskEncryptionParametersDto` object, nothing more

File: environment/src/main/java/com/sequenceiq/environment/parameters/v1/converter/AwsEnvironmentParametersConverter.java [hunk 6]
```diff
@@ -54,6 +59,9 @@ protected void postConvertToDto(Builder builder, BaseParameters source) {
                 .withDynamoDbTableCreation(awsParameters.getS3guardTableCreation())
                 .withFreeIpaSpotPercentage(awsParameters.getFreeIpaSpotPercentage())
                 .withFreeIpaSpotMaxPrice(awsParameters.getFreeIpaSpotMaxPrice())
+                .withAwsDiskEncryptionParameters(AwsDiskEncryptionParametersDto.builder()
+                .withEncryptionKeyArn(awsParameters.getEncryptionKeyArn())
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/parameters/v1/converter/AwsEnvironmentParametersConverter.java:63]: wrong indent

File: environment/src/main/resources/schema/app/20210722121419_CB-13110_Add_the_encryption_Arn_parameter_passed_from_CLI_to_AWS_env_creation_in_Cloudbreak.sql [hunk 7]
```diff
@@ -0,0 +1,8 @@
+-- // CB-13110:Add the encryption Arn parameter passed from CLI to AWS env creation in Cloudbreak
+-- Migration SQL that makes the change goes here.
+ALTER TABLE environment_parameters ADD COLUMN IF NOT EXISTS encryption_key_arn text;
```
Comment (bergerdenes) [inline environment/src/main/resources/schema/app/20210722121419_CB-13110_Add_the_encryption_Arn_parameter_passed_from_CLI_to_AWS_env_creation_in_Cloudbreak.sql:3]: Varchar would be a better type here. Please check for length we use for other ARNs (or CRNs)

File: environment/src/main/java/com/sequenceiq/environment/parameters/dao/domain/AwsParameters.java [hunk 8]
```diff
@@ -12,7 +18,7 @@
 
 @Entity
 @DiscriminatorValue("AWS")
-public class AwsParameters extends BaseParameters {
+public class AwsParameters extends BaseParameters implements AccountIdAwareResource {
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/parameters/dao/domain/AwsParameters.java:21]: is this change really needed?

File: environment/src/main/java/com/sequenceiq/environment/parameters/validation/validators/parameter/AwsParameterValidator.java [hunk 9]
```diff
@@ -93,4 +109,23 @@ private LocationAwareCredential getLocationAwareCredential(EnvironmentDto enviro
                 .withLocation(environment.getLocation().getName())
                 .build();
     }
+
+    //CHECKSTYLE:OFF:FallThroughCheck
+    private ValidationResult validateEntitlement(ValidationResultBuilder validationResultBuilder,
+            AwsDiskEncryptionParametersDto awsDiskEncryptionParametersDto
+             , String accountId) {
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/parameters/validation/validators/parameter/AwsParameterValidator.java:116]: wrong formatting

File: environment/src/main/java/com/sequenceiq/environment/parameters/validation/validators/parameter/AwsParameterValidator.java [hunk 10]
```diff
@@ -93,4 +109,23 @@ private LocationAwareCredential getLocationAwareCredential(EnvironmentDto enviro
                 .withLocation(environment.getLocation().getName())
                 .build();
     }
+
+    //CHECKSTYLE:OFF:FallThroughCheck
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/parameters/validation/validators/parameter/AwsParameterValidator.java:113]: please remove this

File: environment/src/main/java/com/sequenceiq/environment/parameters/validation/validators/parameter/AwsParameterValidator.java [hunk 11]
```diff
@@ -62,6 +67,17 @@ public ValidationResult validate(EnvironmentValidationDto environmentValidationD
                 determineAwsParameters(environmentDto, parametersDto);
             }
         }
+        ValidationResult validationResult;
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/parameters/validation/validators/parameter/AwsParameterValidator.java:70]: Do not create a new validationResult here. Just use the builder.

File: environment/src/test/java/com/sequenceiq/environment/environment/v1/converter/EnvironmentResponseConverterTest.java [hunk 12]
```diff
@@ -317,6 +320,11 @@ private ParametersDto createAwsParameters() {
                 .withAwsParameters(AwsParametersDto.builder()
                         .withDynamoDbTableName("my-table")
                         .build())
+                .withAwsParameters(AwsParametersDto.builder()
+                        .withAwsDiskEncryptionParameters(AwsDiskEncryptionParametersDto.builder()
+                        .withEncryptionKeyArn("dummy-key-arn")
```
Comment (bergerdenes) [inline environment/src/test/java/com/sequenceiq/environment/environment/v1/converter/EnvironmentResponseConverterTest.java:325]: wrong indent

File: environment/src/test/java/com/sequenceiq/environment/parameters/v1/converter/AwsEnvironmentParametersConverterTest.java [hunk 13]
```diff
@@ -52,18 +55,24 @@ void getCloudPlatform() {
...
-                        .withDynamoDbTableCreation(S3GuardTableCreation.CREATE_NEW)
-                        .build())
-                .build();
+        ParametersDto builder = ParametersDto.builder()
+                                .withId(ID)
+                                .withAwsParameters(AwsParametersDto.builder()
+                                    .withDynamoDbTableName(TABLE_NAME)
+                                    .withDynamoDbTableCreation(S3GuardTableCreation.CREATE_NEW)
+                                    .withAwsDiskEncryptionParameters(AwsDiskEncryptionParametersDto.builder()
+                                    .withEncryptionKeyArn("dummy-key-arn")
```
Comment (bergerdenes) [inline environment/src/test/java/com/sequenceiq/environment/parameters/v1/converter/AwsEnvironmentParametersConverterTest.java:64]: You do not use the const in L37

File: environment/src/test/java/com/sequenceiq/environment/parameters/v1/converter/AwsEnvironmentParametersConverterTest.java [hunk 14]
```diff
@@ -52,18 +55,24 @@ void getCloudPlatform() {
...
     void convertTest() {
         when(environmentViewConverter.convert(any(Environment.class))).thenReturn(ENVIRONMENT_VIEW);
-        ParametersDto parameters = ParametersDto.builder()
-                .withId(ID)
-                .withAwsParameters(AwsParametersDto.builder()
-                        .withDynamoDbTableName(TABLE_NAME)
-                        .withDynamoDbTableCreation(S3GuardTableCreation.CREATE_NEW)
-                        .build())
-                .build();
+        ParametersDto builder = ParametersDto.builder()
```
Comment (bergerdenes) [inline environment/src/test/java/com/sequenceiq/environment/parameters/v1/converter/AwsEnvironmentParametersConverterTest.java:58]: the variable should not be renamed as it is not a `builder`

File: environment/src/test/java/com/sequenceiq/environment/parameters/validation/validators/AwsParameterValidatorTest.java [hunk 15]
```diff
@@ -4,13 +4,15 @@
...
 import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 import java.util.stream.Stream;
 
+import org.junit.Assert;
```
Comment (bergerdenes) [inline environment/src/test/java/com/sequenceiq/environment/parameters/validation/validators/AwsParameterValidatorTest.java:15]: please do not mix Junit4 and 5

**Review by bergerdenes — CHANGES_REQUESTED** [review]

File: environment/src/main/java/com/sequenceiq/environment/parameters/dao/domain/AwsParameters.java (hunk 8 above)
Comment (satyabollineni) [reply environment/src/main/java/com/sequenceiq/environment/parameters/dao/domain/AwsParameters.java:21]: Yes, This is needed, Without this the Environment creation is failing with the error "AwsParameters must be a subclass of AccountIdAwareResource"

File: structuredevent-model/src/test/java/com/sequenceiq/environment/parameter/dto/AwsParametersDtoTest.java (hunk 5 above)
Comment (satyabollineni) [reply structuredevent-model/src/test/java/com/sequenceiq/environment/parameter/dto/AwsParametersDtoTest.java:6]: I have made the changes to the test case to check just for the object

File: environment/src/main/java/com/sequenceiq/environment/parameters/v1/converter/AwsEnvironmentParametersConverter.java [hunk 16]
```diff
@@ -54,6 +59,9 @@ protected void postConvertToDto(Builder builder, BaseParameters source) {
                 .withDynamoDbTableCreation(awsParameters.getS3guardTableCreation())
                 .withFreeIpaSpotPercentage(awsParameters.getFreeIpaSpotPercentage())
                 .withFreeIpaSpotMaxPrice(awsParameters.getFreeIpaSpotMaxPrice())
+                .withAwsDiskEncryptionParameters(AwsDiskEncryptionParametersDto.builder()
+                        .withEncryptionKeyArn(awsParameters.getEncryptionKeyArn())
+                .build())
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/parameters/v1/converter/AwsEnvironmentParametersConverter.java:64]: wrong indent

File: structuredevent-model/src/test/java/com/sequenceiq/environment/parameter/dto/AwsParametersDtoTest.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class AwsParametersDtoTest {
+    @Test
+    void testAwsParametersDtoWithEncryptionParameters() {
+        AwsParametersDto dummyAwsParametersDto = AwsParametersDto.builder()
+                .withAwsDiskEncryptionParameters(AwsDiskEncryptionParametersDto.builder()
+                        .withEncryptionKeyArn("dummy-key-arn")
+                .build())
```
Comment (bergerdenes) [inline structuredevent-model/src/test/java/com/sequenceiq/environment/parameter/dto/AwsParametersDtoTest.java:12]: wrong indent

**Review by cegganesh84 — APPROVED** [review]
LGTM

File: environment/src/test/java/com/sequenceiq/environment/environment/v1/converter/EnvironmentApiConverterTest.java [hunk 18]
```diff
@@ -292,6 +293,35 @@ void testInitLoadBalancerDto() {
...
 
+    @Test
+    void testAwsDiskEncryptionParametersAndAwsRequest() {
+        EnvironmentRequest request = createEnvironmentRequest(AWS);
+        request.setAws(AwsEnvironmentParameters.builder()
+                .withAwsDiskEncryptionParameters(
+                        AwsDiskEncryptionParameters.builder()
+                                .withEncryptionKeyArn("dummy-key-arn")
+                                .build())
+                .build());
```
Comment (juhi-09) [inline environment/src/test/java/com/sequenceiq/environment/environment/v1/converter/EnvironmentApiConverterTest.java:304]: `EncryptionKeyArn` gets set as part of `createEnvironmentRequest`. Need not to set it again.

File: environment/src/test/java/com/sequenceiq/environment/parameters/validation/validators/AwsParameterValidatorTest.java [hunk 19]
```diff
@@ -105,6 +112,50 @@ void validateAndDetermineAwsParametersUseExisting(S3GuardTableCreation creation)
...
+
+        ValidationResult validationResult = underTest.validate(environmentValidationDto, environmentDto.getParameters(), ValidationResult.builder());
+
+        assertFalse(validationResult.hasError());
+    }
+
+    @Test
+    public void testWhenEncryptionKeyArnPresentAndEntitlementDisabledThenError() {
+        EnvironmentDto environmentDto = new AwsParameterValidatorTest.EnvironmentDtoBuilder()
+                                        .withAwsParameters(AwsParametersDto.builder()
```
Comment (juhi-09) [inline environment/src/test/java/com/sequenceiq/environment/parameters/validation/validators/AwsParameterValidatorTest.java:130]: Nit - Please keep the indent like in line 118 or this.

File: environment/src/test/java/com/sequenceiq/environment/parameters/validation/validators/AwsParameterValidatorTest.java [hunk 20]
```diff
@@ -105,6 +112,50 @@ void validateAndDetermineAwsParametersUseExisting(S3GuardTableCreation creation)
...
+                                        .withAwsDiskEncryptionParameters(AwsDiskEncryptionParametersDto.builder()
+                                            .withEncryptionKeyArn("dummy-key-arn")
+                                            .build())
+                                            .build())
+                                        .build();
+
+        EnvironmentValidationDto environmentValidationDto = EnvironmentValidationDto.builder().withEnvironmentDto(environmentDto).build();
+        when(entitlementService.isAWSDiskEncryptionWithCMKEnabled(anyString())).thenReturn(false);
+        ValidationResult validationResult = underTest.validate(environmentValidationDto, environmentDto.getParameters(), ValidationResult.builder());
+        assertTrue(validationResult.hasError());
```
Comment (juhi-09) [inline environment/src/test/java/com/sequenceiq/environment/parameters/validation/validators/AwsParameterValidatorTest.java:140]: Assert the error message as well.

**Review by juhi-09 — CHANGES_REQUESTED** [review]

**Review by juhi-09 — APPROVED** [review]
LGTM, Thanks for the fixes

**Review by bergerdenes — APPROVED** [review]
